plugins/providers/alicloud_ram/provider.go

```diff
@@ -0,0 +1,307 @@
+	if err != nil {
+		return err
+	}
+
+	switch g.Resource.Type {
+	case ResourceTypeAccount:
+		switch g.AccountType {
+		case AccountTypeRamUser:
+			username, _, err := splitAliAccountUserId(g.AccountID)
+			if err != nil {
+				return err
+			}
+			for _, permission := range permissions {
+				err = client.GrantAccess(ctx, permission.Name, permission.Type, username)
+				if err != nil && !errors.Is(err, ErrPermissionAlreadyExists) {
+					return err
+				}
+			}
+			return nil
+
+		case AccountTypeRamRole:
+			for _, permission := range permissions {
+				err = client.GrantAccessToRole(ctx, permission.Name, permission.Type, g.AccountID)
+				if err != nil && !errors.Is(err, ErrPermissionAlreadyExists) {
+					return err
+				}
+			}
+			return nil
+
+		default:
+			return ErrInvalidAccountType
+		}
+
+	default:
+		return ErrInvalidResourceType
+	}
+}
+
+func (p *Provider) RevokeAccess(ctx context.Context, pc *domain.ProviderConfig, g domain.Grant) error {
+	client, err := p.getClient(pc)
+	if err != nil {
+		return err
+	}
+
+	permissions, err := getListPermissionsFromGrant(pc, g)
+	if err != nil {
+		return err
+	}
+
+	switch g.Resource.Type {
+	case ResourceTypeAccount:
+		switch g.AccountType {
+		case AccountTypeRamUser:
+			username, _, err := splitAliAccountUserId(g.AccountID)
+			if err != nil {
+				return err
+			}
+			for _, permission := range permissions {
+				err = client.RevokeAccess(ctx, permission.Name, permission.Type, username)
+				if err != nil && !errors.Is(err, ErrPermissionNotExist) {
+					return err
+				}
+			}
+			return nil
+
+		case AccountTypeRamRole:
+			for _, permission := range permissions {
+				err = client.RevokeAccessFromRole(ctx, permission.Name, permission.Type, g.AccountID)
+				if err != nil && !errors.Is(err, ErrPermissionNotExist) {
+					return err
+				}
+			}
+			return nil
+
+		default:
+			return ErrInvalidAccountType
+		}
+
+	default:
+		return ErrInvalidResourceType
+	}
+}
+
+func (p *Provider) GetRoles(pc *domain.ProviderConfig, resourceType string) ([]*domain.Role, error) {
+	return provider.GetRoles(pc, resourceType)
+}
+
+func (p *Provider) GetPermissions(_pc *domain.ProviderConfig, _resourceType, role string) ([]interface{}, error) {
+	return p.PermissionManager.GetPermissions(_pc, _resourceType, role)
+}
+
+func (p *Provider) GetAccountTypes() []string {
+	return getAccountTypes()
+}
+
+func (p *Provider) ListAccess(ctx context.Context, pc domain.ProviderConfig, resources []*domain.Resource) (domain.MapResourceAccess, error) {
+	client, err := p.getClient(&pc)
+	if err != nil {
+		return nil, err
+	}
+
+	return client.ListAccess(ctx, pc, resources)
+}
+
+func (p *Provider) getClient(pc *domain.ProviderConfig) (AliCloudRAMClient, error) {
+	var credentials Credentials
+	err := mapstructure.Decode(pc.Credentials, &credentials)
+	if err != nil {
+		return nil, err
+	}
+
+	providerURN := pc.URN
+	if client, ok := p.Clients[providerURN]; ok && client != nil {
```

Comment:
It seems that in the case of ram_role, there could be a scenario where the stsToken in the client has already expired. However, the current logic only checks the in-memory map and returns the existing client if it is found, without verifying whether the stsToken is still valid. This could lead to errors since the client cannot be used after the token's expiry time,

Reply:
On this case, please check on file `client.go`.
It will generate token on every request:
![image](https://github.com/user-attachments/assets/a84af718-9744-4ece-8824-94b21fdc9303)

Example how we create a request to AliCloud:
![image](https://github.com/user-attachments/assets/1c0055ec-b6f8-4142-9f77-abf1adceaa97)

So I think the expiration will not become a problem with this implementation.

Reply:
yes, just checked this logic we're recreating the client for every new request and only validating client creation in `NewAliCloudRAMClient`.